### DISCUSSION TITLE
Connection to ws://...?format=json&schemaVersion=2.2.0 is failed

### DISCUSSION THREAD

**Discussion by rostamiani** [ObserveRTC/observer#94] [Q&A] [answered]
Hi
I tried to use latest client version (1.3) with latest observer (1.0) and have two problems:
And the connection does not happen!
```
Connection to ws://localhost:7080/samples/myService/my-webrtc-app?format=json&schemaVersion=2.2.0 is failed
```
I found out that Observer `1.0.0` is compatible with Schema `2.0.x` but how can I change the schema version?
I tried to downgrade client to `1.1` and `1.0` but the schema version is still `2.2.0`

## My problems:
- ### Could not find the cause of the error. not on client nor server
- ### Cannot downgrade the schema version of client

**Comment by balazskreith** [comment]
Hi, I will take a look on this soon, thank you for raising it. The server side does not log whats the problem with the session if it has, that's something we need to add.

Has the websocket onerror at the client side send a reason or message? if observer closes it should have some reason

observer:1.0.0 fully compatible with 2.2.0, and 2.1.0 schema, so that's certainly not the issue.

**Comment by balazskreith** [comment]
> Cannot downgrade the schema version of client

Ran to this problem before.
That's a bug or a feature. 

Before client-monitor-js:1.3.0, the package.json had the dependency pointing to schemas:^2.1.0, I think and then it tries the highest version available at the moment of resolving the package. so that one is fixed from 1.3.0 because the package version of the schema is fixed. but like I said compatibility is not the issue between schema:2.2.0 and observer:1.0.0

**Comment by rostamiani** [comment]
Thanks, This is good news... 
But docs says observer 1.0.0 is just compatible with version 2.0.x. This should be fixed
![image](https://user-images.githubusercontent.com/1157108/208644772-e44ff632-e0f3-449e-a7ce-6c57af2aa7be.png)
https://github.com/ObserveRTC/observer#compatibility

**Comment by rostamiani** [comment]
I tried client 1.1.1 and it's working but client 1.3 still disconnects with this error:
```
logger.js:36 ObserveRTC::WebsocketTransport Tue, 20 Dec 2022 10:45:40 GMT Closed by the observer, reason: Not valid or not supported schema version: 2.2.0
ObserveRTC::WebsocketTransport Tue, 20 Dec 2022 10:45:40 GMT Closed
ObserveRTC::Sender Tue, 20 Dec 2022 10:45:43 GMT Error: Failed to send data on an already closed transport
```
And this is the server log:
```
10:45:40.407 [default-nioEventLoopGroup-1-13] INFO  o.o.o.s.SamplesWebsocketController - Session pHIyMrx2aZeHJP9jE3FUHA== is closed
```

**Comment by balazskreith** [comment]
updated the compatibility table https://github.com/ObserveRTC/observer/pull/95 
can you check if you change the format to "protobuf" instead of json it works.
Start checking with json

**Comment by balazskreith** [comment]
I checked the [full-stack-eaxmple](https://github.com/ObserveRTC/full-stack-examples) with docker, and running local but the client-monitor was connected.
schema 2.2.0 should not be accepted with version 1..0.0-beta.
can you check what observer logs out what is the version?
you can always find it after the icon is printed, for example (on my local branch now):
```
          `/o+osso:
              .osss+`  `.-::////::-.`
               `/ssso/osssssssssssssso+:.
                `+sssssso+/:----://ossssso:`
               :ossso/-`            `-/ossso/`
             .ossso-                    -+ssso-
            :ssso:                        -osss:
           -ssso.                          `osss:
          `osso. -::::-` -::::::::-  `-::-. .osso.
          :sss/ .ss+++oo::::/ss/:::.+so++os: :sss/
          +sss- .ss-  /so   .ss.  `os:    `  .ssso
          +sss- .ssoooso.   .ss.  .ss-       .ssso
          /sss: .ss:.-os:   .ss`   :oo/::/+- -sss+
          .ssso``//`  .+:   `//`    `-/++/:` osss-
           /sss+                            /sss+
            +sss+`                        `+sss+`
             /ssso:`                     :osss/`
              .+ssso/.                ./ossso-
                -+sssso+:-.``````.-:+ossss+-
                  `:+osssssssssssssssso+:.
                      .-/++oooooo++/:.`
Observer version: 1.0.1
```

and the last line is the version. I made a new branch https://github.com/ObserveRTC/observer/pull/95 and added warn logs for not accepted samples as well.
you can run that from docker:  ` docker run observertc/observer:1.0.0-SNAPSHOT.version-1.0.1 -p 7080:7080`

**Comment by rostamiani** [comment]
Oh, That was my mistake. I had Beta version inside docker compose. Changed it to 1.0.0 but it didn't fix.
I replaced your image in docker-compose.yml and did compose down and up again. ‌But no warnings logged at the server:thinking: 
(I guess that your image in the previous post is wrong. It should be `1,0,1-SNAPSHOT`)

This is my docker-compose:
```yml
  observer:
    restart: always
    image: observertc/observer:1.0.1-SNAPSHOT.version-1.0.1 
    ports:
      - 7080:7080
    environment:
      - MICRONAUT_CONFIG_FILES=/my-config.yaml
    volumes:
      - ./observer-config/config.yaml:/my-config.yaml

```

**Reply by balazskreith** [reply]
There is also a list of supported schema versions when the observer started and prints it to the console. Can you check that list? 
When a websocket session is established there is a provided schema. The provided schema version should be in the list the observer support. If that is not the case then its an issue we need to open. In the issue the logs from the observer, the websocket address line and the console logs should be there.

I can take a look on this after Xmas

**Comment by balazskreith** [comment]
@rostamiani , please try run observer with command: 
```
docker run -p 7080:7080 observertc/observer:1.0.1-SNAPSHOT.version-1.0.1
```

there is a log line:

> [main] INFO  o.o.observer.sources.SampleSources - Supported schema versions: 2.2.0, 2.1.9, 2.1.8, 2.1.7, 2.1.6, 2.1.5, 2.1.4, 2.1.3, 2.1.2, 2.1.1, 2.1.0

tells the supported schema versions.
When a session is established there is also a log line:
> INFO  o.o.o.s.SamplesWebsocketController - Session U8Twtjt5TpoeFpS5pUWfxQ== is opened, providedSchemaVersion: 2.2.0, providedFormat: json

Or something like this.
If something does not match there should be a log line something like:

> WARN  o.o.o.s.SamplesWebsocketController - Error occurred while creating acceptor for session FnoHPj0xaBodmCJ3lmk8GQ==, providedSchemaVersion: 1.2.3, providedFormat: json. Message: Not valid or not supported schema version: 1.2.3. Supported versions: 2.2.0, 2.1.9, 2.1.8, 2.1.7, 2.1.6, 2.1.5, 2.1.4, 2.1.3, 2.1.2, 2.1.1, 2.1.0

Please keep me update how it goes

**Comment by rostamiani** [comment]
I ran your image and my local schema is in the supported list:
local:
```
ObserveRTC::WebsocketTransport Thu, 12 Jan 2023 07:43:04 GMT Connection to ws://79790e505996:7080/samples/myService/my-webrtc-app?format=json&schemaVersion=2.2.0 is failed. Tried: 0
```
Server:
```
Initialized, supported schema versions 
latest
2.2.0
2.1.9
2.1.8
2.1.7
2.1.6
2.1.5
2.1.4
2.1.3
2.1.2
2.1.1
2.1.0
```
But nothing happened in the server log when I tried to connect!
Just some info logs like this:
```
INFO  o.o.o.r.tasks.CleanCallEntities - Completed Clean Call Entities
INFO  o.o.o.r.tasks.CommitAbstract - Committing ended, context: Committing SfuEntities
INFO  o.o.o.r.tasks.CleanSfuEntities - Completed Clean Sfu Entities
INFO  o.o.observer.BackgroundTasksExecutor - Executing task: Exposing Repository Metric
INFO  o.o.observer.BackgroundTasksExecutor - Executing task: Exposing Hamok Metric
INFO  o.o.observer.BackgroundTasksExecutor - Executing task: Exposing Repository Metric
INFO  o.o.observer.BackgroundTasksExecutor - Executing task: Exposing Hamok Metric
```

**Reply by balazskreith** [reply]
Something does not come together. the new version (and now its released as 1.0.1), the log entry for the supported schema versions not line-separated but comma separated and the text is also different:
> [main] INFO o.o.observer.sources.SampleSources - Supported schema versions: 2.2.0, 2.1.9, 2.1.8, 2.1.7, 2.1.6, 2.1.5, 2.1.4, 2.1.3, 2.1.2, 2.1.1, 2.1.0

Can you check if you run the right docker image. Make sure you remove previous image from docker and repull the whole one again.

**Comment by rostamiani** [comment]
I removed all Observer images and downloaded again. Now this is the schema list:

```
Supported schema versions: 2.2.0, 2.1.9, 2.1.8, 2.1.7, 2.1.6, 2.1.5, 2.1.4, 2.1.3, 2.1.2, 2.1.1, 2.1.0
```

But error still exists. ws connection fails and there is no logs at observer!

**Comment by rostamiani** [comment]
Oh, I forgot to send my code :sweat_smile: 
Maybe I'm using it wrong:

```js
let monitor;
function observeRtcStart() {
  monitor = createClientMonitor({
    collectingPeriodInMs: 3000,
    samplingPeriodInMs: 3000,
    sendingPeriodInMs: 3000,
    sampler: {
      roomId: 'room-id-12345',
    },
    sender: {
      websocket: {
        urls: [
          'ws://localhost:7080/samples/myService/my-webrtc-app',
        ]
      },
      format: 'json'
    }
  });

  const mediaDevice = streamingModule.getMediasoupDevice();
  const statsCollector = monitor.collectors.addMediasoupDevice(mediaDevice);
}
```

**Reply by balazskreith** [reply]
your config is correct at least I don't see any problem with it. Try with 'protobuf' instead of 'json' and see if it makes any difference (should not). 
also please try creating a WebSocket from browser console simply run `new Websocket("ws://localhost:7080/samples/myService/my-webrtc-app")` in that case, we can exclude the possibility of client-monitor-js bug.

I see that in another conversation you use the full-stack-example and that one works for you, at least in that case is connected?!

**Reply by rostamiani** [reply]
Finally it started working today... and sadly, I don't know how!
I just changed the back-end server.
Thanks for your support
I will continue this conversation if it happened again, but everything is perfect for now!

**Comment by balazskreith** [comment, marked as answer]
Check your local firewall rules or maybe your VPN was enabled, don't know but it does not seem to be an observer issue but a network connectivity issue.
